### PR TITLE
Bau - Update readme and add overview documentation

### PR DESCRIPTION
**Pull request by JHjava**
- fix typo with licence file
- add licence to readme
- make readme more accurate
- add an 'overview of the proxy node doc' which is reference by the readme. Similar to what we have in the hub.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,16 +1,23 @@
...
+The eIDAS Proxy Node also runs the [verify-service-provider] (https://github.com/alphagov/verify-service-provider) to build Verify SAML AuthnRequests and parse SAML Responses from the Verify Hub.
 
-This stack allows a user journey from a stub EU connector, through the eIDAS Proxy Node, to a stub UK IDP, and back.
 The eIDAS Proxy Node will not perform matching.
 
+## Architectural Descision Records and documentation
+
+We record our architectural decisions in `doc/adr`.
+
+An technical overview of the Proxy Node is available [here](doc/overview.md).
```
Comment (CharlesIC) [inline README.md]:
```suggestion
A technical overview of the Proxy Node is available [here](doc/overview.md).
```

File: README.md [hunk 2]
```diff
@@ -1,16 +1,23 @@
...
 * `proxy-node-translator`
-* `soft-hsm`
+* `eidas-saml-parser`
 * `stub-idp`
+* `stub-connector`
+
+The eIDAS Proxy Node also runs the [verify-service-provider] (https://github.com/alphagov/verify-service-provider) to build Verify SAML AuthnRequests and parse SAML Responses from the Verify Hub.
 
-This stack allows a user journey from a stub EU connector, through the eIDAS Proxy Node, to a stub UK IDP, and back.
 The eIDAS Proxy Node will not perform matching.
```
Comment (CharlesIC) [inline README.md]:
```suggestion
The eIDAS Proxy Node does not perform matching.
```
Just say it in the present tense as this isn't going to change

File: doc/overview.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
+# GOV.UK Verify UK Proxy Node
+
+## Technical overview
+
+The UK Proxy Node will allow GOV.UK Verify to provide identities as part of the eIDAS scheme. This document describes the architecture and behaviour of the UK Proxy Node.
```
Comment (CharlesIC) [inline doc/overview.md]:
```suggestion
The UK Proxy Node allows GOV.UK Verify to provide identities as part of the eIDAS scheme. This document describes the architecture and behaviour of the UK Proxy Node.
```
Can just use the present tense here as well

File: doc/overview.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+### Role of the UK Proxy Node
+As part of the eIDAS scheme, the UK must host a proxy node to allow users access to GOV.UK Verify when in an EU country. As defined in the eIDAS interoperability architecture, the proxy node will act as a front to one or more of a country’s identity schemes. 
+
+The UK Proxy Node will be introduced into the architecture, as a REST-based microservice, to conform with eIDAS specifications. 
+
+GOV.UK Verify uses a SAML message format that is different from the eIDAS specification. The UK Proxy Node will support the eIDAS SAML Profile and will act as an identity provider to EU member state services and as a Relying Party to GOV.UK Verify.
```
Comment (CharlesIC) [inline doc/overview.md]:
```suggestion
GOV.UK Verify uses a SAML message format that is different from the eIDAS specification. The UK Proxy Node supports the eIDAS SAML Profile and acts as an identity provider to EU member state services and as a Relying Party to GOV.UK Verify. 
```

File: doc/overview.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+Part of the UK Proxy Node, the Translator service receives the GOV.UK Verify Hub responses forwarded by the Gateway service. 
+
+The Translator sends the SAML AuthnResponse to the Verify Service Provider (VSP) to perform the SAML validation. The Translator service will generate an eIDAS AuthnResponse based on the attributes it receives from the VSP. The Translator service then encrypts the signed assertion for the connector node and sends the whole eIDAS SAML response for signing by the Signing service.
+
+Once the eIDAS SAML response is signed, the Translator service passes it on to the Gateway service.
+
+### Verify-Service-Provider
+
+The Verify Service Provider (VSP) generates and translates Security Assertion Markup Language (SAML) messages to and from the GOV.UK Verify Hub. 
+The Gateway service uses the VSP to generate SAML requests to send to the Verify Hub. For responses, the Translator service uses the VSP to validate SAML responses and translate it to JSON.
```
Comment (CharlesIC) [inline doc/overview.md]:
```suggestion
The Gateway service uses the VSP to generate SAML requests to send to the Verify Hub. For responses, the Translator service uses the VSP to validate SAML responses and translate them to JSON. 
```

File: doc/overview.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+The Verify Service Provider (VSP) generates and translates Security Assertion Markup Language (SAML) messages to and from the GOV.UK Verify Hub. 
+The Gateway service uses the VSP to generate SAML requests to send to the Verify Hub. For responses, the Translator service uses the VSP to validate SAML responses and translate it to JSON. 
+
+### Signing-Service
+
+This component is a microservice running on an Amazon EC2 instance. The service will use an Amazon hardware security module (HSM) to sign outbound eIDAS SAML messages. 
+
+## Connecting to Countries
+
+There be one instance of the proxy node services and infrastructure isolation per requesting Country connector node. They will share a single hardware security module (HSM) but will each have independent credentials to access the HSM and an independent signing key held by the HSM.
```
Comment (CharlesIC) [inline doc/overview.md]:
```suggestion
We will host one instance of the proxy node services and infrastructure isolation per requesting Country connector node. All instances will share a single hardware security module (HSM) but will each have independent credentials to access the HSM, and an independent signing key held by the HSM.
```

**Comment by JHjava** [comment]
@CharlesIC Thanks. Changes made. Git makes it so easy!

File: doc/overview.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+The Gateway Service sends the SAML AuthnRequest to the Eidas Saml Parser (ESP) service to validate the data and the signature of the request. 
+
+The ESP is the only service which can consume the countries connector metadata. The ESP will extract the Public Encryption Certificate and Destination of the Country from the metadata and send it back to the Gateway along with the RequestID and Issuer which is extracted from the SAML AuthnRequest.
+
+### Translator
+
+Part of the UK Proxy Node, the Translator service receives the GOV.UK Verify Hub responses forwarded by the Gateway service. 
+
+The Translator sends the SAML AuthnResponse to the Verify Service Provider (VSP) to perform the SAML validation. The Translator service will generate an eIDAS AuthnResponse based on the attributes it receives from the VSP. The Translator service then encrypts the signed assertion for the connector node and sends the whole eIDAS SAML response for signing by the Signing service.
```
Comment (vixus0) [inline doc/overview.md]: What's the "Signing Service"?

File: doc/overview.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+The UK Proxy Node will be introduced into the architecture, as a REST-based microservice, to conform with eIDAS specifications. 
+
+GOV.UK Verify uses a SAML message format that is different from the eIDAS specification. The UK Proxy Node supports the eIDAS SAML Profile and acts as an identity provider to EU member state services and as a Relying Party to GOV.UK Verify. 
+
+The role of the UK Proxy Node is to:
+ - convert eIDAS SAML to Verify SAML messages and vice versa
+ - validate all the incoming requests and outgoing responses 
+ - cryptographically sign all outgoing responses to EU services
+
```
Comment (idrop) [inline doc/overview.md]: Can we add to Proxy Node role
 - encrypt the identity assertions sent to the EU service

File: doc/overview.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+SAML responses going to EU member state connector nodes will be signed with keys stored on an HSM. These are keys with self-signed certificates. The same key will be used to sign the proxy node metadata. 
+
+The Verify Service Provider will be responsible for Signing AuthnRequests and Encrypting Responses from the Verify Hub. These will be stored as sealed secrets. 
+
+| 			Key type           | 		Key location        |
+| ---------------------------- | -------------------------- |
+| Metadata signing key         | AWS Key Management System  |
+| AuthnRespone signing key     | AWS Key Management System  |
+| AuthnRequest signing key     | Verify Service Provider    |
```
Comment (idrop) [inline doc/overview.md]: I'm not sure from this how the EU member state will validate the signature created by the CloudHSM signing keys.
Do we publish the CloudHSM signing certs. I might have missed something from above ☝️ though

Comment (JHjava) [reply doc/overview.md]: We publish the signing cert in the Proxy Node Metadata, which they will consume.

File: doc/overview.md (hunk 8 above)
Comment (JHjava) [reply doc/overview.md]: Sure. I'll add that.

File: doc/overview.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+## Security operations monitoring
+
+Security operations within the organisation will monitor transactions and access. Transaction monitoring will involve correlating messages entering with messages exiting the proxy node. Access monitoring will be put in place using Amazon's CloudWatch logging. Logs will be monitored by a Security Operations team to track access to servers, deployments and key rotations.
+
+## UK Proxy Node key management
+
+SAML responses going to EU member state connector nodes will be signed with keys stored on an HSM. These are keys with self-signed certificates. The same key will be used to sign the proxy node metadata. 
+
+The Verify Service Provider will be responsible for Signing AuthnRequests and Encrypting Responses from the Verify Hub. These will be stored as sealed secrets.
```
[the remark this answers is not in the document]
Comment (JHjava) [reply doc/overview.md]: Yep. Will change.

File: doc/overview.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+## UK Proxy Node key management
+
+SAML responses going to EU member state connector nodes will be signed with keys stored on an HSM. These are keys with self-signed certificates. The same key will be used to sign the proxy node metadata. 
+
+The Verify Service Provider will be responsible for Signing AuthnRequests and Encrypting Responses from the Verify Hub. These will be stored as sealed secrets. 
+
+| 			Key type           | 		Key location        |
+| ---------------------------- | -------------------------- |
+| Metadata signing key         | AWS Key Management System  |
```
[the remark this answers is not in the document]
Comment (JHjava) [reply doc/overview.md]: Yep it should be. Will change.

File: doc/overview.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+| AuthnRespone signing key     | AWS Key Management System  |
```
[the remark this answers is not in the document]
Comment (JHjava) [reply doc/overview.md]: Good spot

File: doc/overview.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+| AuthnRequest signing key     | Verify Service Provider    |
```
[the remark this answers is not in the document]
Comment (JHjava) [reply doc/overview.md]: Yes. That would be clearer.

File: doc/overview.md (hunk 7 above)
Comment (JHjava) [reply doc/overview.md]: Was referring to the CloudHSM client. As discussed, going to remove to prevent confusion.

**Pull request merged**